src/auth/src/credentials/internal/sts_exchange.rs

```diff
@@ -0,0 +1,442 @@
+        let mut params: HashMap<&str, String> = HashMap::new();
+
+        params.insert("grant_type", TOKEN_EXCHANGE_GRANT_TYPE.to_string());
+        params.insert("requested_token_type", ACCESS_TOKEN_TYPE.to_string());
+
+        params.insert("subject_token", req.subject_token);
+        params.insert("subject_token_type", req.subject_token_type);
+
+        if !req.scope.is_empty() {
+            params.insert("scope", req.scope.join(" "));
+        }
+
+        if let Some(audience) = req.audience {
+            params.insert("audience", audience);
+        }
+        if let Some(resource) = req.resource {
+            params.insert("resource", resource);
+        }
+        if let Some(actor_token) = req.actor_token {
+            params.insert("actor_token", actor_token);
+        }
+        if let Some(actor_token_type) = req.actor_token_type {
+            params.insert("actor_token_type", actor_token_type);
+        }
+
+        if let Some(options) = req.extra_options {
+            if let Ok(value) = serde_json::to_value(options) {
+                params.insert("options", value.to_string());
+            }
+        }
+
+        self.execute(req.url, req.authentication, req.headers, params)
+            .await
+    }
+
+    /// execute http request and token exchange
+    pub async fn execute(
+        &self,
+        url: String,
+        client_auth: ClientAuthentication,
+        headers: http::HeaderMap,
+        params: HashMap<&str, String>,
+    ) -> Result<TokenResponse> {
+        let mut headers = headers.clone();
+        let mut params = params.clone();
+        println!("[execute] url: {}", url);
+        client_auth.inject_auth(&mut headers, &mut params);
+        println!("[execute] headers: {:?}", headers);
+
+        let res = self
+            .client
+            .post(url)
+            .form(&params)
+            .headers(headers)
+            .send()
+            .await
+            .map_err(|err| {
+                CredentialsError::from_str(false, format!("failed to request token: {}", err))
+            })?;
+
+        let status = res.status();
+        println!("[execute] status: {:?}", status);
+        println!("[execute] response: {:?}", res);
+        if !status.is_success() {
+            return Err(CredentialsError::from_str(
+                false,
+                format!("error exchanging token, failed with status {status}"),
+            ));
+        }
+        let token_res = res
+            .json::<TokenResponse>()
+            .await
+            .map_err(|err| CredentialsError::new(false, err))?;
+        Ok(token_res)
+    }
+}
+
+/// TokenResponse is used to decode the remote server response during
+/// an oauth2 token exchange.
+#[derive(Deserialize, Default, PartialEq, Debug)]
+pub struct TokenResponse {
+    pub access_token: String,
+    pub issued_token_type: String,
+    pub token_type: String,
+    pub expires_in: u64,
+    pub scope: String,
```

Comment:
Should this be a raw string or a list?

Reply:
from what I checked on other implementations and on the RFC, it is returned just as a string, not a list. We can do some transformation and split with spaces and have a list.